tests/components/gree/test_climate.py

```diff
@@ -0,0 +1,783 @@
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    # First update to make the device available
+    state = hass.states.get(ENTITY_ID)
+    assert state.name == "fake-device-1"
+    assert state.state != STATE_UNAVAILABLE
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_HVAC_MODE,
+        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_HVAC_MODE: HVAC_MODE_AUTO},
+        blocking=True,
+    )
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.state == STATE_UNAVAILABLE
+
+
+async def test_send_power_on(hass, discovery, device, mock_now):
+    """Test for sending power on command to the device."""
+    await async_setup_gree(hass)
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_HVAC_MODE,
+        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_HVAC_MODE: HVAC_MODE_AUTO},
+        blocking=True,
+    )
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.state == HVAC_MODE_AUTO
+
+
+async def test_send_power_on_device_timeout(hass, discovery, device, mock_now):
+    """Test for sending power on command to the device with a device timeout."""
+    device().push_state_update.side_effect = DeviceTimeoutError
+
+    await async_setup_gree(hass)
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_HVAC_MODE,
+        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_HVAC_MODE: HVAC_MODE_AUTO},
+        blocking=True,
+    )
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.state == HVAC_MODE_AUTO
+
+
+async def test_send_target_temperature(hass, discovery, device, mock_now):
+    """Test for sending target temperature command to the device."""
+    await async_setup_gree(hass)
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_TEMPERATURE,
+        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_TEMPERATURE: 25.1},
+        blocking=True,
+    )
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.attributes.get(ATTR_TEMPERATURE) == 25
+
+
+async def test_send_target_temperature_device_timeout(
```

Comment:
What is the expected difference in outcomes between this test and the one above? This test appears to set a DeviceTimeoutError, and the previous one does not, but neither seems to update the ATTR_TEMPERATURE.  I believe this could be intentional and/or am i missing a subtle difference, but thought it might be worth asking.

Reply:
Both of them are updating the ATTR_TEMPERATURE in this case.

Gree is a polling device class with an async implementation, so any value set first update the attribute (synchronously, setters should not operate asynchronously), then the next update loop will async try and update the device state, which may timeout. In the event that the device doesn't accept the change, or the change didn't go through (due to the timeout) the attribute will remain set until the next successful poll.

In the case of Gree, it will actually return the current value in the response to the push state, so the correction is fast. If the device times out, there may be a short delay until the next poll.

This code does not attempt to emulate the device behaviour of course. This test only really validates that the exception thrown from the underlying library is properly and robustly handled in the integration.

Reply:
Thanks for the explanation.